### ISSUE TITLE
tablify doesn't work on lldpRemTable

### ISSUE THREAD

**Issue by moepman** [exhuma/puresnmp#68]
Using tablify on the result of a bulkwalk of a lldpRemTable (1.0.8802.1.1.2.1.4.1) returns the right number of elements but the data within is incomplete. Instead of a comple table entry I only get one entry per line (and I have no idea what it is, but it is no where near the complete table).

**Comment by exhuma** [comment]
Interesting.

If the data does not contain anything confidential, can you enable debug logging and add the output of it into this ticket? Preferably with only the `bulkwalk` call? A microscopic script like the following should suffice to generate output for debugging this:

```
import logging
from puresnmp import bulkwalk
logging.basicConfig(level=logging.DEBUG)

bulkwalk(...)
```

Please note that the community will be visible in that output in both the ASCII and HEX part of the logs.


Additionally, if you have an expected output, this would also help a bunch in localising the error. Maybe taking a screenshot from [mibbrowser](http://ireasoning.com/mibbrowser.shtml) which support tables.

The logging output will definitely be the biggest help.

Also, if you can, try the same using a normal `walk` instead of a `bulkwalk` and let me know if that behaves in any way different.

**Comment by moepman** [comment]
puresnmp:
```
>>> import logging
>>> from puresnmp import bulkwalk
>>> logging.basicConfig(level=logging.DEBUG)
>>> result = bulkwalk("172.23.1.31", "public", ["1.0.8802.1.1.2.1.4.1"])
>>> for row in result:                                                  
...     print('%s: %r' % row)                                           
... 
DEBUG:puresnmp.api.raw:Walking on 1 OIDs using _bulkwalk_fetcher(10)
DEBUG:puresnmp.transport:Sending packet to 172.23.1.31:161 (attempt 1/3)
30 2a 02 01 01 04 06 70  75 62 6c 69 63 a5 1d 02   0*.....public...
04 5d de 9e de 02 01 00  02 01 0a 30 0f 30 0d 06   .].........0.0..
09 28 c4 62 01 01 02 01  04 01 05 00               .(.b........
DEBUG:puresnmp.transport:Received packet:
30 82 01 1f 02 01 01 04  06 70 75 62 6c 69 63 a2   0........public.
82 01 10 02 04 5d de 9e  de 02 01 00 02 01 00 30   .....].........0
82 01 00 30 13 06 0e 28  c4 62 01 01 02 01 04 01   ...0...(.b......
01 04 00 0f 02 02 01 05  30 13 06 0e 28 c4 62 01   ........0...(.b.
01 02 01 04 01 01 04 00  17 03 02 01 04 30 17 06   .............0..
0e 28 c4 62 01 01 02 01  04 01 01 05 00 0f 02 04   .(.b............
05 01 ac 17 03 8d 30 18  06 0e 28 c4 62 01 01 02   ......0...(.b...
01 04 01 01 05 00 17 03  04 06 b4 e9 b0 59 02 d0   .............Y..
30 13 06 0e 28 c4 62 01  01 02 01 04 01 01 06 00   0...(.b.........
0f 02 02 01 07 30 13 06  0e 28 c4 62 01 01 02 01   .....0...(.b....
04 01 01 06 00 17 03 02  01 05 30 21 06 0e 28 c4   ..........0!..(.
62 01 01 02 01 04 01 01  07 00 0f 02 04 0f 30 30   b.............00
31 44 34 35 42 36 39 39  32 46 3a 50 31 30 15 06   1D45B6992F:P10..
0e 28 c4 62 01 01 02 01  04 01 01 07 00 17 03 04   .(.b............
03 47 69 30 30 19 06 0e  28 c4 62 01 01 02 01 04   .Gi00...(.b.....
01 01 08 00 0f 02 04 07  53 57 20 50 4f 52 54 30   ........SW PORT0
22 06 0e 28 c4 62 01 01  02 01 04 01 01 08 00 17   "..(.b..........
03 04 10 47 69 67 61 62  69 74 45 74 68 65 72 6e   ...GigabitEthern
65 74 30                                           et0
DEBUG:puresnmp.api.raw:1 of 1 OIDs need to be continued
1.0.8802.1.1.2.1.4.1.1.4.0.15.2: 5
1.0.8802.1.1.2.1.4.1.1.4.0.23.3: 4
1.0.8802.1.1.2.1.4.1.1.5.0.15.2: b'\x01\xac\x17\x03\x8d'
1.0.8802.1.1.2.1.4.1.1.5.0.23.3: b'\xb4\xe9\xb0Y\x02\xd0'
1.0.8802.1.1.2.1.4.1.1.6.0.15.2: 7
1.0.8802.1.1.2.1.4.1.1.6.0.23.3: 5
1.0.8802.1.1.2.1.4.1.1.7.0.15.2: b'001D45B6992F:P1'
1.0.8802.1.1.2.1.4.1.1.7.0.23.3: b'Gi0'
1.0.8802.1.1.2.1.4.1.1.8.0.15.2: b'SW PORT'
1.0.8802.1.1.2.1.4.1.1.8.0.23.3: b'GigabitEthernet0'
DEBUG:puresnmp.transport:Sending packet to 172.23.1.31:161 (attempt 1/3)
30 2f 02 01 01 04 06 70  75 62 6c 69 63 a5 22 02   0/.....public.".
04 5d de 9e de 02 01 00  02 01 0a 30 14 30 12 06   .].........0.0..
0e 28 c4 62 01 01 02 01  04 01 01 08 00 17 03 05   .(.b............
00                                                 .
DEBUG:puresnmp.transport:Received packet:
30 82 02 4f 02 01 01 04  06 70 75 62 6c 69 63 a2   0..O.....public.
82 02 40 02 04 5d de 9e  de 02 01 00 02 01 00 30   ..@..].........0
82 02 30 30 30 06 0e 28  c4 62 01 01 02 01 04 01   ..000..(.b......
01 09 00 0f 02 04 1e 53  45 50 30 30 31 44 34 35   .......SEP001D45
42 36 39 39 32 46 2e 62  69 6e 61 72 79 2e 6b 69   B6992F.binary.ki
74 63 68 65 6e 30 25 06  0e 28 c4 62 01 01 02 01   tchen0%..(.b....
04 01 01 09 00 17 03 04  13 61 70 30 33 2e 62 69   .........ap03.bi
6e 61 72 79 2e 6b 69 74  63 68 65 6e 30 3f 06 0e   nary.kitchen0?..
28 c4 62 01 01 02 01 04  01 01 0a 00 0f 02 04 2d   (.b............-
43 69 73 63 6f 20 49 50  20 50 68 6f 6e 65 20 37   Cisco IP Phone 7
39 36 31 47 2d 47 45 2c  56 32 2c 20 53 49 50 34   961G-GE,V2, SIP4
31 2e 39 2d 34 2d 32 53  52 31 2d 31 53 30 82 01   1.9-4-2SR1-1S0..
01 06 0e 28 c4 62 01 01  02 01 04 01 01 0a 00 17   ...(.b..........
03 04 81 ee 43 69 73 63  6f 20 49 4f 53 20 53 6f   ....Cisco IOS So
66 74 77 61 72 65 2c 20  43 33 35 30 30 20 53 6f   ftware, C3500 So
66 74 77 61 72 65 20 28  41 50 33 47 31 2d 4b 39   ftware (AP3G1-K9
57 37 2d 4d 29 2c 20 56  65 72 73 69 6f 6e 20 31   W7-M), Version 1
35 2e 33 28 33 29 4a 46  35 2c 20 52 45 4c 45 41   5.3(3)JF5, RELEA
53 45 20 53 4f 46 54 57  41 52 45 20 28 66 63 32   SE SOFTWARE (fc2
29 0a 54 65 63 68 6e 69  63 61 6c 20 53 75 70 70   ).Technical Supp
6f 72 74 3a 20 68 74 74  70 3a 2f 2f 77 77 77 2e   ort: http://www.
63 69 73 63 6f 2e 63 6f  6d 2f 74 65 63 68 73 75   cisco.com/techsu
70 70 6f 72 74 0a 43 6f  70 79 72 69 67 68 74 20   pport.Copyright 
28 63 29 20 31 39 38 36  2d 32 30 31 38 20 62 79   (c) 1986-2018 by
20 43 69 73 63 6f 20 53  79 73 74 65 6d 73 2c 20    Cisco Systems, 
49 6e 63 2e 0a 43 6f 6d  70 69 6c 65 64 20 4d 6f   Inc..Compiled Mo
6e 20 32 39 2d 4a 61 6e  2d 31 38 20 32 32 3a 34   n 29-Jan-18 22:4
34 20 62 79 20 70 72 6f  64 5f 72 65 6c 5f 74 65   4 by prod_rel_te
61 6d 30 14 06 0e 28 c4  62 01 01 02 01 04 01 01   am0...(.b.......
0b 00 0f 02 04 02 24 00  30 14 06 0e 28 c4 62 01   ......$.0...(.b.
01 02 01 04 01 01 0b 00  17 03 04 02 20 00 30 14   ............ .0.
06 0e 28 c4 62 01 01 02  01 04 01 01 0c 00 0f 02   ..(.b...........
04 02 24 00 30 14 06 0e  28 c4 62 01 01 02 01 04   ..$.0...(.b.....
01 01 0c 00 17 03 04 02  20 00 30 1b 06 16 28 c4   ........ .0...(.
62 01 01 02 01 04 02 01  03 00 0f 02 01 04 81 2c   b..............,
17 03 81 0d 02 01 01 30  1a 06 15 28 c4 62 01 01   .......0...(.b..
02 01 04 02 01 03 00 17  03 01 04 81 2c 17 01 2b   ............,..+
02 01 03                                           ...
DEBUG:puresnmp.api.raw:0 of 1 OIDs need to be continued
1.0.8802.1.1.2.1.4.1.1.9.0.15.2: b'SEP001D45B6992F.binary.kitchen'
1.0.8802.1.1.2.1.4.1.1.9.0.23.3: b'ap03.binary.kitchen'
1.0.8802.1.1.2.1.4.1.1.10.0.15.2: b'Cisco IP Phone 7961G-GE,V2, SIP41.9-4-2SR1-1S'
1.0.8802.1.1.2.1.4.1.1.10.0.23.3: b'Cisco IOS Software, C3500 Software (AP3G1-K9W7-M), Version 15.3(3)JF5, RELEASE SOFTWARE (fc2)\nTechnical Support: http://www.cisco.com/techsupport\nCopyright (c) 1986-2018 by Cisco Systems, Inc.\nCompiled Mon 29-Jan-18 22:44 by prod_rel_team'
1.0.8802.1.1.2.1.4.1.1.11.0.15.2: b'$\x00'
1.0.8802.1.1.2.1.4.1.1.11.0.23.3: b' \x00'
1.0.8802.1.1.2.1.4.1.1.12.0.15.2: b'$\x00'
1.0.8802.1.1.2.1.4.1.1.12.0.23.3: b' \x00'
>>> 
```

snmptable binary
```
# snmptable -c public -v2c -m LLDP-MIB 172.23.1.31 1.0.8802.1.1.2.1.4.1
SNMP table: LLDP-MIB::lldpRemTable

 lldpRemChassisIdSubtype     lldpRemChassisId lldpRemPortIdSubtype     lldpRemPortId  lldpRemPortDesc                 lldpRemSysName                                                                                                                                                                                                                                 lldpRemSysDesc lldpRemSysCapSupported lldpRemSysCapEnabled
          networkAddress    "01 AC 17 03 8D "                local "001D45B6992F:P1"          SW PORT SEP001D45B6992F.binary.kitchen                                                                                                                                                                                                  Cisco IP Phone 7961G-GE,V2, SIP41.9-4-2SR1-1S               "24 00 "             "24 00 "
              macAddress "B4 E9 B0 59 02 D0 "        interfaceName             "Gi0" GigabitEthernet0            ap03.binary.kitchen Cisco IOS Software, C3500 Software (AP3G1-K9W7-M), Version 15.3(3)JF5, RELEASE SOFTWARE (fc2) Technical Support: http://www.cisco.com/techsupport Copyright (c) 1986-2018 by Cisco Systems, Inc. Compiled Mon 29-Jan-18 22:44 by prod_rel_team               "20 00 "             "20 00 "
```

**Comment by exhuma** [comment]
Awesome thanks. I'll get to this as soon as possible. I am currently working on #66 which, with any luck is actually related to this one. I'm in the process of cleaning up and waiting on confirmation that it works as expected from the submitter of #66 

I might start working on this in parallel in the coming days. Having the dumps will speed this up tremendously.

**Comment by exhuma** [comment]
It turns out that this is actually not a bug. The table you access is rooted at the OID `1.0.8802.1.1.2.1.4.1`, which has 9 nodes, but the returned OIDs have 14 nodes (f.ex. `1.0.8802.1.1.2.1.4.1.1.11.0.23.3`).

In general, you have to pass in the number of the root OID nodes + 1 to tablify. So in this case it should be:

```
varbinds = bulkwalk(...)
table = tablify(varbinds, 10)
```

This information is normally available in the MIBs which are not parsed by `puresnmp`. I beleive, a signature of `tablify` which takes the root-OID would be much easier and I will consider supporting that in the future. That way, the code would look like this:

```
varbinds = bulkwalk(..., '1.0.8802.1.1.2.1.4.1', ...)
table = tablify(varbinds, '1.0.8802.1.1.2.1.4.1')
```

I am aware that this is non-intuitive, and [the cookbook example](https://puresnmp.readthedocs.io/en/latest/cookbook/table_with_bulk.html) does not make any mention of this.

The API docs for `tablify` are also not very clear.

I will close this issue for now, and open a new one to update the documentation.

**Comment by moepman** [comment]
> This information is normally available in the MIBs which are not parsed by `puresnmp`. I beleive, a signature of `tablify` which takes the root-OID would be much easier and I will consider supporting that in the future. That way, the code would look like this:
> 
> ```
> varbinds = bulkwalk(..., '1.0.8802.1.1.2.1.4.1', ...)
> table = tablify(varbinds, '1.0.8802.1.1.2.1.4.1')
> ```

From the cookbook I assumed that tablify would just find out that all items in the result set have the same prefix. Good to know that (for now) I have to pass (and figure out) that number for each of the tables I want to to have a look at.

**Comment by exhuma** [comment]
Yeah, I realise that this is not quite convenient. A lot of different design decisions went into that, and I might revisit some of that in the near future. For some transparency, here are some of my thoughts I had about this in the past:

* Most of the code was written with `asyncio` in mind from the start even though it was not supported until later. But the main focus around that was to isolate the network I/O as much as possible. This has the advantage that some results (especially in "walk" and "bulkwalk" operations) can be processed in the client application as soon as they arrive from the network. To process does not need to wait for the walk to finish completely. *However*, this has the downside that "table" operations don't necessarily see enough context to automatcially determine the root OID. Especially for large tables. Passing in the context information into that function allows the piecemal processing again. There is however still some logic in the code which currently forces me to fetch more than I want to for tables. So while the initial aim is still valid, the current implementation is sub-optimal.
* The `puresnmp.table` function is in a wonderful position to abstract the "base-oid" detection as it already abstracts the combination of `walk` and `tablify`. Additionally, I did not implement a `bulktable(...)` function because I wanted to keep the numbers of exposed functions to a minimum. Considering the usefulness of `bulktable(...)` this was a really silly decision. And next to the new issue (#69), it would be really useful to offer a `bulktable()` function, and then determine the base-oid inside that function instead of `tablify`.

Discussions like this are really helpful for a library like this, so I thank you for bringing this to my attention again. It's always great to see things from a different angle. 😃

And I hope you are able to solve the problem with the `num_base_nodes` argument even if it is a bit redundant at the moment.
